### ISSUE TITLE
use of wrong "C" instance produces hard-to-diagnose failure

### ISSUE THREAD

**Issue by rogpeppe** [go-check/check#41]
I was recently trying to find out what was going wrong in a failed
test. The symptom was that the tests failed saying "49 PASSED 7 FAILED"
but with no other error message (and there were only 49 tests in total).
Running with the race detector enabled showed that there was a race.

The reason was that a test suite was setting up an helper object in
SetUpTest that held on to the *C value passed in SetUpTest, and then in
one of the tests the helper object did an assert on it.

This is obviously a wrong thing to do, but it's an easy mistake to make
and it would have saved me lots of time if the failure mode had been
less obscure - for example if the assert had panicked because the C
object had already been discarded.

Here's some code that reproduces the issue.
http://play.golang.org/p/3CXBOBWLB_ I've been testing using go tip
(+2db587c)

For the record, the culprit was NewFakeAPI in
github.com/juju/juju/worker/provisioner.

**Comment by howbazaar** [comment]
Yeah, I found out that the *C passed into the tests was not the same *C that was passed into the setup test and tear down test methods when I think they probably should be.

**Comment by niemeyer** [comment]
It's intentional that SetUpTest's *C remains usable after the method returns. That's why logic inside that method can set up logging to happen via its *C that will remain valid for every test in the suite.

It's not clear what you are reporting here, though. That's not a bug. Is there another bug you are trying to report?

**Comment by rogpeppe** [comment]
Do you agree that the code above should not demonstrate a race condition?

**Comment by rogpeppe** [comment]
BTW I'm not complaining that the *C from SetUpTest remains usable after the method
returns, which seems fine, but that _if_ you use it to do a check you a) don't see any errors
printed and b) incur a race condition.

**Comment by cellfish** [comment]
I believe #22 is a duplicate of this issue (or vice versa). I also have been bitten when assuming that checks on the *C passed during setup would fail a test. My usage scenario has been the use of fakes. During test startup some fakes are created capturing the *C to do a number of asserts during the test run.

I think the "problem" (i.e. source for confusion) i that it is a *C rather than C being passed to the different test methods. Even if a C was passed any content (like log output etc) could be references inside so that different copies of C would still share certain things.
